### PR TITLE
Updates Gateway API Design Doc

### PR DESCRIPTION
**Pull request by danehans**
Updates the Gateway API design doc based on the operator managing Gateway objects.

xref: https://docs.google.com/document/d/1fq_zSozkPeImD1SoBvxQ6_mXmtqfGCn_77UQgFn0mAY/edit?usp=sharing

Signed-off-by: Daneyon Hansen <daneyonhansen@gmail.com>

### CODE REVIEW

File: design/gateway-apis-implementation.md
Comment (skriss) [inline design/gateway-apis-implementation.md:131]: is there a corresponding Condition for this?

File: design/gateway-apis-implementation.md
Comment (skriss) [inline design/gateway-apis-implementation.md:132]: Is there a Condition for this?

File: design/gateway-apis-implementation.md
Comment (skriss) [inline design/gateway-apis-implementation.md:147]: Is this an exhaustive list of all the Conditions + Reasons which should cause Contour to *not* try to program a Listener/its routes? And otherwise, it *should* try to program that listener/routes?

File: design/gateway-apis-implementation.md
Comment (danehans) [reply design/gateway-apis-implementation.md:297]: @skriss I'm trying to generalize the design to use "gateway controller" since it appears that the maintainer team has not achieved consensus on where the Gateway controller should live. A potential approach to unblock us, is to support multiple [controller](https://github.com/kubernetes-sigs/gateway-api/blob/v0.2.0/apis/v1alpha1/gatewayclass_types.go#L49-L63) domain/paths. 

__Option 1- Operator Managed:__ The operator will own a Gateway of class "projectcontour.io/contour-operator" and be responsible for managing the Gateway, including status. Contour will be responsible reading Gateway status to manage HTTPRoutes and TLSRoutes, including status. 

__Option 2- Contour Managed:__ The operator will __not__ manage a Gateway of class "projectcontour.io/contour". Instead, Contour will be responsible for managing the Gateway, including status. The GatewayClass should either not reference an object or reference an object other than the Contour custom resource... maybe a ConfigMap. With this approach, users are expected to manage Contour using standard Kubernetes manifests. Contour will need to ensure the following for a Gateway:
- If [addresses](https://github.com/kubernetes-sigs/gateway-api/blob/v0.2.0/apis/v1alpha1/gateway_types.go#L115-L132) are specified, ensure they match Envoy service addresses and surface the appropriate Gateway status condition if not.
- Ensure each [listener port/protocol](https://github.com/kubernetes-sigs/gateway-api/blob/v0.2.0/apis/v1alpha1/gateway_types.go#L163-L184) matches the port name/protocol in the Envoy service and surface the appropriate Gateway/Listener status condition if not.

Supporting both approaches will be a short-term approach that allows us to figure out the model that works best for Contour. Thoughts? cc: @stevesloka @sunjayBhatia @youngnick

File: design/gateway-apis-implementation.md
Comment (danehans) [reply design/gateway-apis-implementation.md:157]: Yes, [RouteGatewayStatus](https://github.com/kubernetes-sigs/gateway-api/blob/v0.2.0/apis/v1alpha1/shared_types.go#L178-L195) should always be managed by Contour using condition type "Admitted". When "Admitted=False", Contour should handle the conflict according to the [guidelines](https://gateway-api.sigs.k8s.io/guidelines/). The gateway controller will watch HTTPRoutes/TLSRoutes and update Gateway/Listener status accordingly.

File: design/gateway-apis-implementation.md
Comment (danehans) [reply design/gateway-apis-implementation.md:147]: The update intends to cover the different Listener condition types and reasons. Let me know if I missed anything.

**Comment by danehans** [comment]
@skriss commit `775c201` addresses the above resolved feedback.

**Comment by danehans** [comment]
@sunjayBhatia commit `f43c572` updates the PR based on your feedback above, PTAL.

File: design/gateway-apis-implementation.md
Comment (youngnick) [reply design/gateway-apis-implementation.md:326]: While I agree that having something that doesn't require installing two things would be preferable, we need to not be pausing development on Gateway support while we figure out what that looks like. What we have in this current design is deliverable with what we currently have, and I would rather have something working first before spending a lot of engineering effort refactoring.

**Review by youngnick — APPROVED** [review]
This LGTM, as a place to get what we have right now working. I agree that we should discuss further options, but any of the options that I've seen discussed so far are multiple-month engineering efforts, and think having something working at all is better than nothing.

**Comment by stevesloka** [comment]
> This LGTM, as a place to get what we have right now working. I agree that we should discuss further options, but any of the options that I've seen discussed so far are multiple-month engineering efforts, and think having something working at all is better than nothing.

I disagree, now is the time to figure these things out while the api is alpha and there are no users. If @danehans, @skriss, & @sunjayBhatia feel the same the that's fine, but if we don't do it now, it won't happen ever.

**Comment by youngnick** [comment]
I agree that this should not be merged unless and until we have the whole team in agreement that this is what we are going to do for the long term.

**Comment by danehans** [comment]
Setting as "Draft" as we work through [the revised implementation plan](https://docs.google.com/document/d/1kOrGqHXULYO6YLxGCOtVyjz_7mfcW7Yi1WXCpQz6_Qo/edit?usp=sharing).

**Pull request closed without merging**
